activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -360,9 +363,9 @@ def clear_reloadable_connections!
       # Raises:
       # - ActiveRecord::ConnectionTimeoutError no connection can be obtained from the pool.
       def checkout(checkout_timeout = @checkout_timeout)
-        connection = checkout_and_verify(acquire_connection(checkout_timeout))
-        connection.lock_thread = @lock_thread
```

Comment:
Losing this [without replacement] would be bad, though presumably it just needs to move to `pin_connection!`

Reply:
How so? AFAIK it's private API, not meant to be used.

Reply:
The effect of this change is still that multiple threads see the same adapter instance; without this line the adapter doesn't know it needs to enable locking.

Reply:
But it doesn't need to, because only the pinned connection is returned and it's always locked.

Reply:
Ohhhh sorry, I was looking in the adapter, didn't realise the connection locking had moved to fixtures.

That does feel a bit far away (given that pinning a non-locked connection = possible segfault, that's a lot of latitude to hand to a caller), but does not have the problem I was imagining. :+1:

Reply:
Yeah, I might actually move the whole thing inside the pool (add lock, open transaction, pin connection).